src/critnib/critnib.c

```diff
@@ -116,14 +116,27 @@ struct critnib_node {
 struct critnib_leaf {
     word key;
     void *value;
+    /*
+     * When cb_free_leaf() is set and values of leaves should be freed,
+     * critnib_remove() sets to_be_freed to the value of the leaf
+     * and sets value to NULL, because the value of a removed leaf
+     * is not valid anymore and should not be returned to a user.
+     * The value of the leaf saved in to_be_freed is freed in critnib_release()
+     * when the reference count of the leaf drops to 0.
+     */
+    void *to_be_freed;
```

Comment:
i do not see a reason/usecase to have two separate ptrs to keep the same data.  Are you sure that we need two ptrs?

Reply:
The reason is described above in the comment. If value is set it will be returned to a user. We have to move it somewhere else to be able to free it when the reference counter drops to 0. Yes, I am sure.

Reply:
Done